controllers/hyperparametertuningjob/hyperparametertuningjob_controller.go

```diff
@@ -19,437 +19,292 @@ package hyperparametertuningjob
-	ctx.Job.ObjectMeta.Finalizers = append(ctx.Job.ObjectMeta.Finalizers, SageMakerResourceFinalizerName)
-	ctx.Log.Info("Add finalizer and Requeue")
-	prevGeneration := ctx.Job.ObjectMeta.GetGeneration()
-	if err := r.Update(ctx, &ctx.Job); err != nil {
-		ctx.Log.Error(err, "Failed to add finalizer", "StatusUpdateError", ctx)
-		return RequeueIfError(err)
-	}
+	switch ctx.TuningJobDescription.HyperParameterTuningJobStatus {
+	case sagemaker.HyperParameterTuningJobStatusInProgress:
+		ctx.HPOTrainingJobSpawner.SpawnMissingTrainingJobs(ctx, *ctx.TuningJob)
 
-	return RequeueImmediatelyUnlessGenerationChanged(prevGeneration, ctx.Job.ObjectMeta.GetGeneration())
-}
+		if controllers.HasDeletionTimestamp(ctx.TuningJob.ObjectMeta) {
+			// Request to stop the job
+			if _, err := ctx.SageMakerClient.StopHyperParameterTuningJob(ctx, ctx.TuningJobName); err != nil && !clientwrapper.IsStopHyperParameterTuningJob404Error(err) {
+				return r.updateStatusAndReturnError(ctx, ReconcilingTuningJobStatus, errors.Wrap(err, "Unable to delete hyperparameter tuning job"))
+			}
+			// Describe the new state of the job
+			if ctx.TuningJobDescription, err = ctx.SageMakerClient.DescribeHyperParameterTuningJob(ctx, ctx.TuningJobName); err != nil {
+				return r.updateStatusAndReturnError(ctx, ReconcilingTuningJobStatus, errors.Wrap(err, "Unable to describe SageMaker hyperparameter tuning job"))
+			}
+		}
+		break
 
-func (r *HyperparameterTuningJobReconciler) getHyperParameterTuningJobName(state hpojobv1.HyperparameterTuningJob) string {
-	return GetGeneratedJobName(state.ObjectMeta.GetUID(), state.ObjectMeta.GetName(), 32)
-}
+	case sagemaker.HyperParameterTuningJobStatusCompleted:
+		if err = r.addBestTrainingJobToStatus(ctx); err != nil {
+			return r.updateStatusAndReturnError(ctx, ReconcilingTuningJobStatus, errors.Wrap(err, "Unable to add best training job to status"))
+		}
+		fallthrough
 
-func (r *HyperparameterTuningJobReconciler) getSageMakerDescription(ctx reconcileRequestContext) (*sagemaker.DescribeHyperParameterTuningJobOutput, awserr.RequestFailure) {
-	describeRequest := ctx.SageMakerClient.DescribeHyperParameterTuningJobRequest(&sagemaker.DescribeHyperParameterTuningJobInput{
-		HyperParameterTuningJobName: ctx.Job.Spec.HyperParameterTuningJobName,
-	})
-
-	describeResponse, describeError := describeRequest.Send(ctx)
-	log := ctx.Log.WithName("getSageMakerDescription")
-
-	if awsErr, requestFailed := describeError.(awserr.RequestFailure); requestFailed {
-		if r.isSageMaker404Response(awsErr) {
-			log.Info("Job does not exist in sagemaker")
-			return nil, nil
-		} else {
-			log.Info("Non-404 error response from DescribeHyperparameterTuningJob")
-			return nil, awsErr
+	case sagemaker.HyperParameterTuningJobStatusStopped, sagemaker.HyperParameterTuningJobStatusFailed:
+		if controllers.HasDeletionTimestamp(ctx.TuningJob.ObjectMeta) {
+			return r.cleanupAndRemoveFinalizer(ctx)
```

Comment:
Why is this duplicated? Line 168 does this, this should noop and let 168 handle cleanup.

Reply:
Line 168 only acts if there is no description available. This handles the case where we have already successfully stopped the SageMaker resource. They are both cleaned up in the same way, however.

Reply:
I see what you mean. The paginator for spawned training jobs would return a 404 if we attempt to clean up a job that's not in SageMaker. Good catch, thank you.